### PR TITLE
Add DK underwriting guidelines

### PR DESCRIPTION
**Pull request by mightyost**
This PR adds the underwriting guidelines for Denmark:

**Regular (Home Content, Accident, Travel)**

- Apartment size: 5- 250 sqm
- Age: 18-∞ years
- Household size: 1-7 people insured total (6 co-insured)

**Student (Home Content, Accident, Travel)**

- Apartment size: 5 - 100 sqm
- Age: 18 - 30 years
- Household size: 1-2 people insured total (1 co-insured)

**ALL** 

- Validating ssn
- Matching birthdate with ssn

To make it easier to add rules having dependencies on each other and also simplifying validation of a field supporting different errors, the rule framework was refactored. Old implementation required rules to provide a static error code and a `validate` function. The validation function should return true if rule breached with the single static error code.

With this PR the `validate` function will now return the error code (if any, else null) to raise if there are any problems.

# Jira Issue: [IPC-6] 

## What?
- Add DK underwriting guidelines

## Why?
- Support for DK

## Optional checklist
- [x] Codescouted
- [x] Unit tests written
- [x] Tested locally

[IPC-6]: https://hedvig.atlassian.net/browse/IPC-6

### CODE REVIEW

File: src/main/kotlin/com/hedvig/underwriter/service/guidelines/DanishAccidentGuidelines.kt [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    override fun validate(data: DanishAccidentData): BreachedGuidelineCode? {
+        if (data.coInsured < 0) {
+            return NEGATIVE_NUMBER_OF_CO_INSURED
+        }
+
+        if (data.isStudent && data.coInsured > 1) {
+            return TOO_HIGH_NUMBER_OF_CO_INSURED
+        }
+
+        if (!data.isStudent && data.coInsured > 6) {
```
Comment (Michael-Duivestein) [inline src/main/kotlin/com/hedvig/underwriter/service/guidelines/DanishAccidentGuidelines.kt:26]: We should consider moving these hardcoded values to some sort of config or DB at some point in time.

File: src/main/kotlin/com/hedvig/underwriter/service/guidelines/BreachedGuidelinesCodes.kt [hunk 2]
```diff
@@ -3,6 +3,8 @@ package com.hedvig.underwriter.service.guidelines
 typealias BreachedGuidelineCode = String
 
 object BreachedGuidelinesCodes {
+    val OK = null
```
Comment (Michael-Duivestein) [inline src/main/kotlin/com/hedvig/underwriter/service/guidelines/BreachedGuidelinesCodes.kt:6]: I would feel a lot more comfortable with not using null as a legitimate code. Is there a special reason we're using null?

File: src/main/kotlin/com/hedvig/underwriter/service/guidelines/SwedishApartmentGuidelines.kt [hunk 3]
```diff
@@ -15,66 +15,56 @@ import java.time.temporal.ChronoUnit
...
+    override fun validate(data: SwedishApartmentData): BreachedGuidelineCode? {
+        if (data.livingSpace!! < 1) {
+            return TOO_SMALL_LIVING_SPACE
         }
-}
 
-object SwedishStudentApartmentLivingSpaceNotMoreThan50Sqm : BaseGuideline<SwedishApartmentData> {
-    override val breachedGuideline = STUDENT_TOO_MUCH_LIVING_SPACE
+        if (!data.isStudent && data.livingSpace > 250) {
+            return TOO_MUCH_LIVING_SPACE
```
Comment (AlexSJ1990) [inline src/main/kotlin/com/hedvig/underwriter/service/guidelines/SwedishApartmentGuidelines.kt:51]: maybe TOO_LARGE_LIVING_SPACE

**Review by AlexSJ1990 — COMMENTED** [review]
Looks good!

File: src/main/kotlin/com/hedvig/underwriter/model/QuoteData.kt [hunk 4]
```diff
@@ -187,7 +187,11 @@ data class DanishHomeContentsData(
 
     override fun updateName(firstName: String, lastName: String) = this.copy(firstName = firstName, lastName = lastName)
     override fun updateEmail(email: String) = this.copy(email = email)
-    override fun updateSsn(ssn: String) = this.copy(ssn = ssn)
+    override fun updateSsn(ssn: String) =
+        if (ssn.isValidDanishSsn())
+            this.copy(ssn = ssn)
+        else
+            throw IllegalArgumentException("Invalid SSN")
```
[the remark this answers is not in the document]
Comment (mightyost) [reply src/main/kotlin/com/hedvig/underwriter/model/QuoteData.kt]: Fixed.

File: src/main/kotlin/com/hedvig/underwriter/service/guidelines/DanishAccidentGuidelines.kt (hunk 1 above)
Comment (mightyost) [reply src/main/kotlin/com/hedvig/underwriter/service/guidelines/DanishAccidentGuidelines.kt:26]: Yes, a lot of things should be more configuration driven...

File: src/main/kotlin/com/hedvig/underwriter/service/guidelines/BreachedGuidelinesCodes.kt (hunk 2 above)
Comment (mightyost) [reply src/main/kotlin/com/hedvig/underwriter/service/guidelines/BreachedGuidelinesCodes.kt]: Fixed!

File: src/main/kotlin/com/hedvig/underwriter/service/guidelines/DanishAccidentGuidelines.kt [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+        return OK
+    }
+}
+
+object DanisAccidentStudentAgeGuideline : BaseGuideline<DanishAccidentData> {
+
+    override fun validate(data: DanishAccidentData): BreachedGuidelineCode? {
+
+        if (data.isStudent && data.age() !in 18..30) {
```
[the remark this answers is not in the document]
Comment (mightyost) [reply src/main/kotlin/com/hedvig/underwriter/service/guidelines/DanishAccidentGuidelines.kt:38]: Yes, student: 18-30 yrs according to Jira task.

File: src/main/kotlin/com/hedvig/underwriter/service/guidelines/SwedishApartmentGuidelines.kt (hunk 3 above)
Comment (mightyost) [reply src/main/kotlin/com/hedvig/underwriter/service/guidelines/SwedishApartmentGuidelines.kt:51]: Agree but do not dare to change this. The value is written to db and I do not know if it is safe to change it?

**Review by Michael-Duivestein — APPROVED** [review]
Nice!

File: src/main/kotlin/com/hedvig/underwriter/service/guidelines/SwedishApartmentGuidelines.kt (hunk 3 above)
Comment (AlexSJ1990) [reply src/main/kotlin/com/hedvig/underwriter/service/guidelines/SwedishApartmentGuidelines.kt:51]: ah sorry hadn't realised!

File: src/main/kotlin/com/hedvig/underwriter/service/guidelines/DanishAccidentGuidelines.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (AlexSJ1990) [reply src/main/kotlin/com/hedvig/underwriter/service/guidelines/DanishAccidentGuidelines.kt:38]: 👍

**Review by AlexSJ1990 — APPROVED** [review]

**Pull request merged**